### PR TITLE
feature/worker/permission syncer: perforce: sync HOST field using IP addresses

### PR DESCRIPTION
**Pull request by ggilmore**
The perforce permission syncer has been adapted to now read and save the HOST field from the perforce protections table, which contains the IP address(es) that the path rule that it applies to.  It uses the updated sub_repository_rules store methods introduced in https://github.com/sourcegraph/sourcegraph/pull/63811/.

### Notes

- There is some existing logic in the permissions syncer that attempts to re-insert the existing sub_repo_permissions if we encounter a temporary (timeout, etc.) error when syncing. However, there is an edge case: what do we do if the existing permissions don't have an IP address associated with them yet (they were inserted before the updated permission syncer ran)? For simplicity, in this PR I leaned towawrds correctness - I fail the operation outright (I'd rather temporarirly lock someone out rather than accidentally leak information). I implemented a more robust straetgy for this in https://github.com/sourcegraph/sourcegraph/pull/64086. 

## Test plan

- The existing unit tests have been adapted to use the new authz.SubRepoPermissionsWithIP structs (I use wildcard IP addresses). 

- The big new test to pay attention to is TestScanIPPermissions (and the associated `sample-protects-ip.txt` file). 

## Changelog

- The perforce permissions syncer has been updated to save the IP address associated with each sub_repository_permissions rule.

### CODE REVIEW

File: internal/authz/providers/perforce/protects_test.go [hunk 1]
```diff
@@ -296,6 +296,98 @@ func TestScanFullRepoPermissions(t *testing.T) {
 	}
 }
 
+func TestScanIPPermissions(t *testing.T) {
```
Comment (ggilmore) [inline internal/authz/providers/perforce/protects_test.go:299]: This is the main test to pay attention to, as it tests non-wildcard IP addresses.

File: internal/authz/providers/perforce/protects_test.go [hunk 2]
```diff
@@ -296,6 +296,98 @@ func TestScanFullRepoPermissions(t *testing.T) {
...
+		"//depot/local/",
+		"//depot/test/",
+		"//depot/rickroll/",
+		"//not-depot/not-main/", // no rules exist
+	}
+
+	perms := &authz.ExternalUserPermissions{
+		SubRepoPermissions: make(map[extsvc.RepoID]*authz.SubRepoPermissionsWithIPs),
+	}
+	if err := scanProtects(logger, testParseP4ProtectsRaw(t, rc), fullRepoPermsScanner(logger, perms, p.depots), false); err != nil {
```
Comment (ggilmore) [inline internal/authz/providers/perforce/protects_test.go:329]: Note that this test https://linear.app/sourcegraph/issue/SRC-516/truncate-sub-repository-permissions-table-when-toggling

File: cmd/worker/internal/authz/perms_syncer.go [hunk 3]
```diff
@@ -532,11 +532,16 @@ func (s *permsSyncerImpl) fetchUserPermsViaExternalAccounts(ctx context.Context,
 				extPerms = new(authz.ExternalUserPermissions)
 
 				// Load last synced sub-repo perms for this user and provider.
-				currentSubRepoPerms, err := s.db.SubRepoPerms().GetByUserAndService(ctx, user.ID, provider.ServiceType(), provider.ServiceID())
+				// Note:
+				// When fetching the xisting sub-repo perms, we can't afford to backfill any IP entries with "*".
+				// Since this data is later saved, these entries with NULL IPs will get saved
+				// as explicit wildcard entries in the database (which isn't the same and might result in information leakage).
+				// @ggilmore Thinks that we can't trade correctness for availability here, but I'm open to suggestions.
+				currentSubRepoPerms, err := s.db.SubRepoPerms().GetByUserAndServiceWithIPs(ctx, user.ID, provider.ServiceType(), provider.ServiceID(), false) // TODO@ggilmore: Should we backfill here? This actually bad since we eventually save these permissions again, where we might
```
Comment (eseliger) [inline cmd/worker/internal/authz/perms_syncer.go:540]: where we might..?

File: cmd/worker/internal/authz/perms_syncer.go [hunk 4]
```diff
@@ -532,11 +532,16 @@ func (s *permsSyncerImpl) fetchUserPermsViaExternalAccounts(ctx context.Context,
 				extPerms = new(authz.ExternalUserPermissions)
 
 				// Load last synced sub-repo perms for this user and provider.
-				currentSubRepoPerms, err := s.db.SubRepoPerms().GetByUserAndService(ctx, user.ID, provider.ServiceType(), provider.ServiceID())
+				// Note:
+				// When fetching the xisting sub-repo perms, we can't afford to backfill any IP entries with "*".
+				// Since this data is later saved, these entries with NULL IPs will get saved
+				// as explicit wildcard entries in the database (which isn't the same and might result in information leakage).
+				// @ggilmore Thinks that we can't trade correctness for availability here, but I'm open to suggestions.
+				currentSubRepoPerms, err := s.db.SubRepoPerms().GetByUserAndServiceWithIPs(ctx, user.ID, provider.ServiceType(), provider.ServiceID(), false)
```
Comment (eseliger) [inline cmd/worker/internal/authz/perms_syncer.go:540]: This means dropping all subrepo perms after an upgrade, regardless of if the IP restrictions mode is on?

Comment (ggilmore) [reply cmd/worker/internal/authz/perms_syncer.go:540]: @eseliger Yes, for this specific PR. However, see  https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64086/feat-worker-permission-syncing-make-sub-repo-permissions-re-insertion-fall-back-to-original-paths-if-ips-not-added-yet

**Review by eseliger — APPROVED** [review]

**Review by pjlast — APPROVED** [review]

**Comment by ggilmore** [comment]
### Merge activity

* **Jul 30, 10:40 AM EDT**: @ggilmore started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/64010).

**Pull request merged**